File: solutions/problem_2049_count_nodes_with_the_highest_score/solution.cpp
/*
 * Problem 2049: Count Nodes With the Highest Score
 * ==============================================
 * Difficulty: Medium
 * Tags: Array, Tree, Depth-First Search, Binary Tree
 * Pattern: DFS Tree Traversal
 *
 * Time Complexity:  O(n)
 * Space Complexity: O(h)
 */

#include <algorithm>
#include <functional>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    int countHighestScoreNodes(vector<int>& parents) {
        // DFS on binary tree - O(n) time, O(h) space
        function<int(TreeNode*)> dfs = [&](TreeNode* node) -> int {
            if (!node) return 0;
            int left = dfs(node->left);
            int right = dfs(node->right);
            return 1 + max(left, right);
        };
        return dfs(parents);
    }
};

File: solutions/problem_2049_count_nodes_with_the_highest_score/solution.py
"""
Problem 2049: Count Nodes With the Highest Score
==============================================
Difficulty: Medium
Tags: Array, Tree, Depth-First Search, Binary Tree
Pattern: DFS Tree Traversal

Time Complexity:  O(n)
Space Complexity: O(h)
"""

class Solution:
    def countHighestScoreNodes(self, parents: List[int]) -> int:
        # DFS on binary tree - O(n) time, O(h) space
        def dfs(node):
            if not node:
                return 0
            left = dfs(node.left)
            right = dfs(node.right)
            return 1 + max(left, right)
        
        result = dfs(parents)
        return result
